File: 01-Programming-Languages/C/Algorithms/String/palindrome.c
/*
Problem: Palindrome Check
Check if a given string is a palindrome (reads same forwards and backwards).

Approach:
- Two-pointer technique from both ends

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: "racecar"
Output: 1 (true)
*/

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool is_palindrome(char str[])
{
    int left = 0, right = strlen(str) - 1;
    while (left < right)
        if (str[left++] != str[right--])
            return false;
    return true;
}

int main()
{
    char str[] = "racecar";
    printf("%s -> %s\n", str, is_palindrome(str) ? "true" : "false");
    return 0;
}

File: 01-Programming-Languages/PHP/Algorithms/String/palindrome.php
<?php

/*
Problem: Valid Palindrome
Description: Check if a string reads the same forwards and backwards.

Approach:
- Use two pointers, one from start and one from end
- Compare characters while moving inward
- If all match, it's a palindrome

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input:  "racecar"
Output: true

Input:  "hello"
Output: false
*/

function isPalindrome(string $s): bool {
    $left = 0;
    $right = strlen($s) - 1;
    while ($left < $right) {
        if ($s[$left] !== $s[$right]) return false;
        $left++;
        $right--;
    }
    return true;
}

$tests = ["racecar", "hello", "madam", "php"];
foreach ($tests as $s) {
    $res = isPalindrome($s) ? "true" : "false";
    echo "\"$s\" -> $res\n";
}
